File: src/harness_core/dispatch/plugin_registry.py
"""Phase 7: PluginRegistry — register, discover, validate, and search plugins."""

from __future__ import annotations

import json
import threading
from pathlib import Path

from .plugin_system import (
    ALL_KNOWN_PERMISSIONS,
    VALID_TRUST_LEVELS,
    PluginManifest,
    _parse_manifest,
)


PLUGIN_REGISTRY_SCHEMA_VERSION = "plugin_registry.v1"


class PluginRegistry:
    """In-memory registry for plugin manifests with discovery and validation."""

    def __init__(self) -> None:
        self._registered: dict[str, PluginManifest] = {}
        self._lock = threading.Lock()

    def register_plugin(self, manifest: PluginManifest) -> bool:
        errors = self.validate_manifest(manifest)
        if errors:
            return False
        if manifest.plugin_id in self._registered:
            return False
        self._registered[manifest.plugin_id] = manifest
        return True

    def unregister_plugin(self, plugin_id: str) -> bool:
        if plugin_id in self._registered:
            del self._registered[plugin_id]
            return True
        return False

    def discover_plugins(self, plugin_dir: str | Path) -> list[PluginManifest]:
        plugin_dir = Path(plugin_dir)
        if not plugin_dir.exists() or not plugin_dir.is_dir():
            return []

        manifests: list[PluginManifest] = []
        for path in sorted(plugin_dir.glob("*.json")):
            try:
                with open(path, "r") as f:
                    raw = json.load(f)
                manifest = _parse_manifest(raw)
                errors = self.validate_manifest(manifest)
                if not errors:
                    manifests.append(manifest)
            except (json.JSONDecodeError, KeyError):
                continue
        return manifests

    def validate_manifest(self, manifest: PluginManifest) -> list[str]:
        errors: list[str] = []

        if not manifest.plugin_id:
            errors.append("plugin_id is required")
        if not manifest.name:
            errors.append("name is required")
        if not manifest.version:
            errors.append("version is required")
        if not manifest.author:
            errors.append("author is required")

        if manifest.trust_level not in VALID_TRUST_LEVELS:
            errors.append(f"invalid trust_level: '{manifest.trust_level}'")

        for perm in manifest.permissions:
            if perm not in ALL_KNOWN_PERMISSIONS:
                errors.append(f"unknown permission '{perm}'")

        return errors

    def list_registered(self) -> list[PluginManifest]:
        return list(self._registered.values())

    def get_plugin(self, plugin_id: str) -> PluginManifest | None:
        return self._registered.get(plugin_id)

    def search_plugins(self, query: str) -> list[PluginManifest]:
        query_lower = query.lower()
        results: list[PluginManifest] = []
        for manifest in self._registered.values():
            if query_lower in manifest.name.lower() or query_lower in manifest.author.lower():
                results.append(manifest)
        return results

File: engine/src/infrastructure/plugin_registry.rs
use super::plugin_system::{
    validate_manifest_fields, PluginManifest, ALL_KNOWN_PERMISSIONS, VALID_TRUST_LEVELS,
};
use std::collections::HashMap;

pub struct PluginRegistry {
    registered: HashMap<String, PluginManifest>,
}

impl Default for PluginRegistry {
    fn default() -> Self {
        Self::new()
    }
}

impl PluginRegistry {
    pub fn new() -> Self {
        Self {
            registered: HashMap::new(),
        }
    }

    pub fn register_plugin(&mut self, manifest: &PluginManifest) -> bool {
        let errors = validate_manifest_fields(manifest);
        if !errors.is_empty() {
            return false;
        }
        if self.registered.contains_key(&manifest.plugin_id) {
            return false;
        }
        self.registered
            .insert(manifest.plugin_id.clone(), manifest.clone());
        true
    }

    pub fn unregister_plugin(&mut self, plugin_id: &str) -> bool {
        self.registered.remove(plugin_id).is_some()
    }

    pub fn validate_manifest(&self, manifest: &PluginManifest) -> Vec<String> {
        let mut errors = Vec::new();
        if manifest.plugin_id.is_empty() {
            errors.push("plugin_id is required".to_string());
        }
        if manifest.name.is_empty() {
            errors.push("name is required".to_string());
        }
        if manifest.version.is_empty() {
            errors.push("version is required".to_string());
        }
        if manifest.author.is_empty() {
            errors.push("author is required".to_string());
        }
        if !VALID_TRUST_LEVELS.contains(&manifest.trust_level.as_str()) {
            errors.push(format!("invalid trust_level: '{}'", manifest.trust_level));
        }
        for perm in &manifest.permissions {
            if !ALL_KNOWN_PERMISSIONS.contains(&perm.as_str()) {
                errors.push(format!("unknown permission '{}'", perm));
            }
        }
        errors
    }

    pub fn list_registered(&self) -> Vec<&PluginManifest> {
        self.registered.values().collect()
    }

    pub fn get_plugin(&self, plugin_id: &str) -> Option<&PluginManifest> {
        self.registered.get(plugin_id)
    }

    pub fn search_plugins(&self, query: &str) -> Vec<&PluginManifest> {
        let query_lower = query.to_lowercase();
        self.registered
            .values()
            .filter(|m| {
                m.name.to_lowercase().contains(&query_lower)
                    || m.author.to_lowercase().contains(&query_lower)
            })
            .collect()
    }
}
